bip-0326.mediawiki

```diff
@@ -0,0 +1,116 @@
+
+Absolute locktimes are also still used, so we should keep using nLockTime, but also often use nSequence.
+
+== Specifications ==
+
+When wallets create transactions spending UTXOs protected by BIP341 taproot, they should set either an nLockTime value or nSequence values to discourage fee sniping, by allowing the transaction to only be mined in the next block after the tip, not the current block. This BIP suggests 50% probability for using nLockTime and 50% for nSequence. If nSequence is set it should apply to at least one of the inputs of the transaction, if it has multiple inputs. It is suggested that on-chain wallets pick an input randomly.
+
+Wallets should also have a second random branch which sets the nLockTime or nSequence value even further back, so that transactions that are delayed after signing for whatever reason (e.g. high-latency mix networks) have better privacy. Existing behaviour is that with a probability of 10%, choose a random number between 0 and 99, and subtract it from the current block height. See the Bitcoin Core and Electrum source codes linked in the references for an example.
+
+nSequence can only encode up to a max of 65535 for the block distance, see BIP68[5], so if the UTXOs being spent have more confirmations than that then the wallet should use nLockTime instead.
+
+=== Pseudocode ===
+
+<source>
+def apply_anti_fee_sniping_fields(transaction, rbf_set):
+    # bip68 requires v=2
+    transaction.version = 2
+    # Initialize all nsequence to indicate the requested RBF state
+    # nsequence can not be 2**32 - 1 in order for nlocktime to take effect
+    for input in transaction.inputs:
+        if rbf_set:
+            input.nsequence = 2**32 - 3
+        else:
+            input.nsequence = 2**32 - 2
+    # always set nlocktime if any of the transaction inputs have more
+    # confirmations than 65535 or are not taproot inputs, or have
+    # unconfirmed inputs
+    # otherwise choose either nlocktime or nsequence with 50% probability
+    if any(map(lambda input: input.confirmations() > 65535
+            || !input.is_taproot() || input.confirmations() == 0,
+            transaction.inputs)) || randint(2) == 0:
+        transaction.nlocktime = blockchain.height()
+        if randint(10) == 0:
+            transaction.nlocktime = max(0, transaction.nlocktime
+            - randint(0, 99))
+        # nsequence must be set in order for nlocktime to take effect
+    else:
+        transaction.nlocktime = 0
+        input_index = randint(len(transaction.inputs))
+        transaction.inputs[input_index].nsequence = transaction.inputs\
+            [input_index].confirmations()
+        if randint(10) == 0:
+            transaction.inputs[input_index].nsequence = max(0,
+                transaction.inputs[input_index].nsequence - randint(0, 99))
```

Comment:
This might again drop the nSequence below `1`. Can be fixed by

```py
nseq -= randint(0, nseq)
nseq -= randint(0, nseq - 1)
```

Depending on whether randint is inclusive or exclusive.


```suggestion
            transaction.inputs[input_index].nsequence -= randint(0, transaction.inputs[input_index].nsequence))
```

Reply:
What do you think of `max(1, transaction.inputs[input_index].nsequence - randint(0, 99))` ?

i.e. Changing `max(0,` into `max(1,` ?

The max stops the nseq going below 1. It also doesnt make assumptions about randint being exclusive or inclusive

Reply:
Jup, that is fine, too.

Side-note: Some wallet software that is written in lower languages might use an unsigned int type for nsequence, in which case subtraction isn't well defined. But I guess they can figure out a way that works.